### PR TITLE
[REFACTOR] 모델 변경 사항 반영

### LINKED ISSUE #24: [REFACTOR] 모델 변경 사항 반영
## 📄 설명

- 기존의 모델에 대한 성능이 낮게 나와 딥러닝 팀에서 개선한 새로운 모델 pt파일을 받아 재배포하겠습니다
- 테스트 및 성능 개선을 위한 dlq에 대한 처리도 보완하겠습니다

---

## ✅ 작업할 내용
---

### PR DESCRIPTION
**Pull request by youyeon11**
## 📌 작업 목적

- 모델 변경을 반영하여 반환타입 리팩토링 및 새로운 라이브러리 추가

---

## 🗂 작업 유형
- [ ] 기능 추가 (Feature)
- [ ] 버그 수정 (Bug Fix)
- [x] 리팩터링 (Refactor)
- [ ] 성능 개선 (Performance)
- [ ] 테스트 추가/수정 (Test)
- [x] 빌드/배포/환경 설정 (Chore)
- [ ] 문서 수정 (Docs)

---

## 🔨 주요 작업 내용

- timm 라이브러리가 추가되었습니다.
```txt
timm
```

- model 변경 (LigthCNN -> EfficientNet)
```python
    def _load_model(self, model_path: Path) -> EfficientNetBaseline:
        import __main__
        __main__.EfficientNetBaseline = EfficientNetBaseline

        object = torch.load(model_path, map_location=self.device, weights_only=False)

        if isinstance(object, nn.Module) :
            # 그 자체로 모델일 때
            model = object.to(self.device)
        elif isinstance(object, dict) :
            # 반환 타입이 state_dict
            state_dict = object
            for k in ['state_dict', 'model_state_dict', 'model']:
                if k in object and isinstance(object[k], dict):
                    state_dict[k] = object[k]
                    break

            model = EfficientNetBaseline(self.num_classes).to(self.device)

            missing, unexpected = model.load_state_dict(state_dict, strict=False)
            if missing or unexpected:
                print(f"[load_state_dict] missing keys: {missing}, unexpected keys: {unexpected}")
        else:
            # type 일치하지 않음
            raise TypeError(f"Unsupported checkpoint type: {type(object)}")

        model.eval()
        return model
```
- 반환 타입이 모델 자체일 때랑 dict 형식일 때가 있어서 케이스 분리

---

## 🧪 테스트 결과

image

---

## 📎 관련 이슈
- Closes #24 
---

### CODE REVIEW

**Comment by youyeon11** [comment]
## 🤖 Gemini 코드 리뷰 결과

## 1. 주요 변경 사항 요약 및 의도 파악

이 PR은 모델을 LightCNN에서 EfficientNet으로 변경하고 이에 따른 코드 수정 및 라이브러리 추가를 포함하는 리팩토링 작업입니다.  모델 변경으로 인해 `torch.load` 함수의 반환 타입이 모델 자체이거나 `state_dict` 형태일 수 있어 이를 처리하는 로직이 추가되었습니다.  또한, 새로운 모델 EfficientNet을 사용하기 위해 `timm` 라이브러리가 추가되었고, 모델 파일도 변경되었습니다.  이러한 변경을 통해 모델 성능 향상 및 코드 개선을 목표로 하는 것으로 이해됩니다.  관련 이슈 #24를 해결하는 PR입니다.


## 2. 코드 품질 및 가독성

* **코드 스타일:** 전반적으로 PEP 8 스타일 가이드를 잘 준수하고 있습니다. 다만,  `if isinstance(object, nn.Module):` 와 `elif isinstance(object, dict):` 사이에 빈 줄이 없는 부분 등 일부 간격 조정이 필요해 보입니다.
* **변수/함수명:**  변수명 `object`는 너무 일반적이라서 구체적인 의미를 나타내도록 `loaded_model` 또는 `checkpoint` 와 같이 변경하는 것이 좋겠습니다.  `state_dict` 변수명은 명확하고 적절합니다.
* **주석/문서화:**  `_load_model` 함수 내부의 주석은 각 조건문의 의도를 잘 설명하고 있어서 이해하기 쉽습니다.  하지만, 함수 자체에 대한 docstring이 없으므로 추가하는 것이 좋겠습니다.  `EfficientNetBaseline` 클래스에도 docstring을 추가하여  `pretrained`, `dropout` 매개변수에 대한 설명을 포함하는 것이 좋습니다.
* **중복 코드:**  현재 눈에 띄는 중복 코드는 없습니다.


## 3. 잠재적 버그 및 엣지 케이스

* **논리적 오류:** `state_dict` 를 로드하는 부분에서  `'state_dict'`, `'model_state_dict'`, `'model'` 키 중 하나라도 존재하면 해당 값을 사용하도록 되어 있습니다. 만약 여러 키가 동시에 존재한다면 어떤 키가 우선순위를 갖는지 명확하지 않아 예상치 못한 결과가 발생할 수 있습니다. 우선순위를 명시적으로 정의하거나, 여러 키가 존재하는 경우 예외를 발생시키는 것이 좋겠습니다.
* **경쟁 상태 (Race Condition):** 현재 코드에서는 멀티쓰레드/비동기 환경에서 발생할 수 있는 경쟁 상태는 보이지 않습니다.
* **에러 핸들링:** `_load_model` 함수에서 `TypeError`를 발생시켜 지원하지 않는 checkpoint 타입을 처리하고 있는 것은 좋습니다.  하지만 `load_state_dict`에서 `missing` 또는 `unexpected` 키가 발생하는 경우, 현재는 단순히 출력만 하고 있습니다.  이러한 상황을 어떻게 처리할지 (예: 에러 로깅, 특정 기본값 사용)에 대한 명확한 정책이 필요합니다.


## 4. 성능 및 효율성

* **시간 복잡도:** 모델 로딩 부분에서  `for k in ['state_dict', 'model_state_dict', 'model']:` 루프는 최대 3번만 반복하므로 성능에 큰 영향을 미치지 않을 것으로 예상됩니다.  전반적인 성능은  EfficientNet 모델 자체의 성능에 더 의존적일 것입니다.
* **자원 사용:**  현재 코드에서 메모리 누수나 불필요한 I/O, 과도한 DB 쿼리 등이 발생할 가능성은 보이지 않습니다.
* **최적화 제안:**  `state_dict` 로딩 로직에서 여러 키 중 하나를 선택하는 부분을 개선하여  우선순위를 명확히 하고 예외 처리를 추가하여 더욱 안전하게 구현할 수 있습니다.


## 5. 보안 및 아키텍처

* **보안 취약점:**  현재 코드에서 특별한 보안 취약점은 보이지 않습니다.
* **아키텍처 적합성:**  모델 변경은 애플리케이션의 아키텍처에 영향을 미칠 수 있는 중요한 변경입니다.  이 PR이 기존 시스템 아키텍처와 어떻게 부합하는지에 대한 설명이 PR 본문에 추가되면 좋겠습니다.  예를 들어,  모델 변경으로 인해 다른 부분의 코드 수정이 필요한지,  API 스펙에 변경이 있는지 등을 명시하는 것이 좋습니다.
* **확장성:**  모델 변경은 향후 다른 모델로의 변경을 용이하게 하기 위해  추상화 계층을 도입하는 것을 고려해 볼 수 있습니다.  예를 들어,  `_load_model` 함수가 특정 모델 클래스에 의존하지 않고  더 일반적인 인터페이스를 사용하도록 리팩토링하면  새로운 모델을 추가할 때  `_load_model` 함수를 수정하지 않고도  새로운 모델 클래스만 구현하면 되도록 변경할 수 있습니다.
